.github/CONTRIBUTING.md

```diff
@@ -74,6 +74,44 @@ Avoid nesting selectors, this will make it easier to maintain in the future.
 }
 ```
 
+### Variable naming
+
+Sass variables should be named according to the
+`$component-state-property-variant` format for consistency. Examples:
```

Comment:
One comment on the naming convention. I brought this up when we talked last week. At least for me – other people are welcomed to disagree – when I name something I try to start large and work my way down the line in a parent/child relationship. For me personally, this reads a little backward. A variant, for example, is a child of a component and not a child of a variant, but it also has both states and properties.

Here's what I'd suggest changing it to.
```scss
$component-variant-state-property
```

```scss
// component: button
// variant: ghost
// state: hover
// property: color
$button-ghost-hover-color: #fff;
```

If a token doesn't require a variant or state then you can omit them.

I also think this helps to quickly differentiate two similar tokens side by side.

```scss
$button-hover-color-danger: #fff;
$button-hover-color-ghost: #fff;
// vs.
$button-danger-padding: #fff;
$button-ghost-padding: #fff;
```

Maybe it's just me though, 🤷‍♂️ so take it for what it is.😄

Reply:
@jeffchew yeah – assuming expressive is a variant of a button – that's what I'm suggesting in ways of naming.

Reply:
@photodow I don't have a strong preference here. `$component-variant-state-property` works for me. We never had this naming convention defined so the code base currently has examples like:

```scss
  /// @type Number
  /// @access public
  /// @group button
  /// Uses the same padding-y as normal buttons, but removes extra padding-right
  $button-padding-ghost: calc(0.875rem - 3px) 12px !default !global;
```

But it's not the end of the world if some of our existing, legacy, variables aren't 100% to the new naming. We could alias it with a new variable and mark it for future deprecation, or just roll with it.

Reply:
Personally, having only a single hyphen as a deliminator makes it really hard to read these variables. And as you get to more complex components I think it's going to be even more difficult (eg. `data-table-header-cell-static-background-hover` is `component-child-property-state` plus add in the fact that this component has two-word parts making it even more indistinguishable).  I know for class names WH doesn't follow BEM but uses SUITCSS (to intentionally be different from Carbon) but SUITCSS doesn't define things like variables and mixins so this is kind of what we came up with: https://pages.github.ibm.com/Watson-Health/docs/getting-started/developers/naming-conventions.html#sass-examples

The general idea is:
* a `namespace-Component--modifier_state-child-property`
* for two-word values, we use camel case (or pascal case for the Component) so that the hyphens act as a true separator.

Reply:
@elizabethsjudd I was thinking through what constraints custom properties give us, as a `$namespace-Component--modifier_state-child-property` Sass variable could be represented as a `--namespace-Component--modifier_state-child-property` custom property. As I understand it that's valid syntax (custom properties are case sensitive).

A different constraint could be how we are tending to author styling in JS for future design asset generation, or CSS-in-JS approaches, etc. Doing so starts as the camel cased JS `inverseSupport01` and then our build puts it in kebab-cased Sass `$inverse-support-01`. Too rigid of a naming pattern could make that difficult.

Looking to external references, Bootstrap hasn't diverged from kebab-cased variables after all these years. https://github.com/twbs/bootstrap-sass/blob/master/assets/stylesheets/bootstrap/_variables.scss

Reply:
That's fair, I'm relatively need to CSS-in-JS so I definitely understand wanting to have something compatible and certainly correct me if I'm doing something that's not supported.
However, the main difference I see between bootstrap and Carbon is that we have a lot more variants/modifiers that also have states which makes for must longer/compound variable names. The hyphen works in bootstrap because all of their variables are short and simple making them relatively scan-able (though I do think they could benefit from different deliminators as well).  The depth and complexity of Carbon is what makes me concerned with going with a hyphen-only approach. Musts IMO are definitely using as few and shorter words as possible and namespacing to reduce conflicts. But, I still think there are some things that can be considered to have things work to make the more compound variable names scan-able, for instance:
- different case considerations
- use of all allowed variable syntaxes in JS with a "map" between the CSS/Sass versions

Taking the example (`namespace-Component--modifier_state-child-property`) of a ghost button's icon hover color:
* CSS: `--bx-Button--ghost_hover-icon-color`
* Sass: `$bx-Button--ghost_hover-icon-color`
* JS: `bxButton$ghost_hoverIconColor` 
     * personally, i feel like the `_` in JS is more impactful then then `$` so I would maybe even opt for something like: `bxButton_ghost$hoverIconColor`

or really considering the JS format move the state to the end (`namespace-Component--modifier-child-property_state`):
* CSS: `--bx-Button--ghost-icon-color_hover`
* Sass: `$bx-Button--ghost-icon-color_hover`
* JS: `bxButton_ghostIconColor$hover`